### PR TITLE
[CALCITE-6488] Ensure collations created by RelCollations are canonized once

### PR DESCRIPTION
**Pull request by rubenada**
More info: CALCITE-6488

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/RelCollations.java [hunk 1]
```diff
@@ -270,7 +270,7 @@ public static RelCollation shift(RelCollation collation, int offset) {
     for (RelFieldCollation fc : collation.getFieldCollations()) {
       fieldCollations.add(fc.shift(offset));
     }
-    return new RelCollationImpl(fieldCollations.build());
+    return RelCollationTraitDef.INSTANCE.canonize(new RelCollationImpl(fieldCollations.build()));
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/rel/RelCollations.java:273]: Looks good, maybe some JavaDoc could be added here

File: core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java [hunk 2]
```diff
@@ -96,9 +96,7 @@ public SortJoinCopyRule(Class<? extends Sort> sortClass,
     if (leftFieldCollation.isEmpty()) {
       newLeftInput = join.getLeft();
     } else {
-      final RelCollation leftCollation =
-          RelCollationTraitDef.INSTANCE.canonize(
-              RelCollations.of(leftFieldCollation));
+      final RelCollation leftCollation = RelCollations.of(leftFieldCollation);
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java:99]: One question, why did you change it? I observed that you only changed the shift interface, and the shift interface was not called here because this part itself is redundant.

**Review by caicancai — APPROVED** [review]
LGTM

File: core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java (hunk 2 above)
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java:99]: [RelCollations.of](https://github.com/apache/calcite/pull/3874/files#diff-d50ce55582ce0f63226670b6234a635b70884a7055fce75df03043b4a77f9f62L85) is already canonizing, only shift needs to be changed to avoid the need of canonization in the callers (this is explained in details in the Jira ticket), so it looks good to me

**Review by asolimando — APPROVED** [review]
LGTM

File: core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java (hunk 2 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/rel/rules/SortJoinCopyRule.java:99]: Thanks for your answer, I also noticed this when I looked at the code

**Review by NobiGo — APPROVED** [review]

**Comment by rubenada** [comment]
Thanks @caicancai , @asolimando , @NobiGo  for the review!
